### ISSUE TITLE
Remove the dependency on openssl

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#7888]
Seems like we only use it for SSL through hyper. I feel like it should be possible to do better here. Maybe @briansmith has ideas.

**Comment by Ms2ger** [comment]
CC @seanmonstar @reem

**Comment by seanmonstar** [comment]
I can help with the "make it work through hyper" part. Pieces to do:

In the `Cargo.toml`s:
```toml
[dependencies]
hyper = { version = "0.6", default-features = "false" }
```

In the `net` crate (I'd suggest in a new file `ssl.rs`):

```rust
use hyper;
use hyper::net::{Ssl, HttpStream};

pub struct ServoSsl {
    context: SomeSortOfContextThingyProbablyMaybe
}

pub struct ServoSslStream {
    inner: HttpStream,
    ssl: SomeThingFromContextMaybe
}

// impl {Clone, io::Read, io::Write, hyper::net::NetworkStream} for ServoSslStream

impl hyper::Ssl for ServoSsl {
    type Stream = ServoSslStream;
    fn wrap_client(&self, stream: HttpStream, host: &str) -> hyper::Result<Self::Stream> {
        let ssl = try!(self.context_with_host_or_something(host));
        Ok(ServoSslStream {
            inner: stream,
            ssl: ssl
        })
    }

    // this is used when using `hyper::Server::https()`.
    fn wrap_server(&self, _: HttpStream) -> hyper::Result<Self::Stream> {
        unimplemented!()
    }
}
```

And finally, use this `Ssl` with the `HttpsConnector` in `http_loader.rs`:

```rust
pub fn create_http_connector() -> Arc<Pool<Connector>> {
    let connector = HttpsConnector::new(ServoSsl {
        context: make_context_or_something()
    });

    Arc::new(Pool::with_connector(Default::default(), connector))
}
```

**Comment by briansmith** [comment]
> Seems like we only use it for SSL through hyper. I feel like it should be possible to do better here. Maybe @briansmith has ideas.

I already discussed this with Alex and Patrick. I'm still waiting to hear back from Mozilla.

**Comment by drbo** [comment]
Mozilla has its own library (NSS) &ndash; maybe this could be used?

**Comment by jdm** [comment]
We would need to write or find Rust bindings to it, first.

**Comment by Ms2ger** [comment]
Using NSS might be a possibility, and it certainly hasn't had as much bad press as openssl, but fundamentally it's still a lot of C attack surface.

**Comment by Ms2ger** [comment]
CC @alex

**Comment by Gaelan** [comment]
How much work would it be to port NSS to Rust? It is hard to gauge how large NSS is because it bundles libraries with its own code.

**Comment by alex** [comment]
Port it, or write bindings?

**Comment by Gaelan** [comment]
Port it. This resolves the "C attack surface" problem. There would also be a good chance of finding problems in the the original code that we could contribute upstream.

**Comment by frewsxcv** [comment]
One might have an easier experience porting things incrementally for something like [ring](https://github.com/briansmith/ring) (based off BoringSSL) since it's written in pure C, whereas NSS is writtenin C++ and the C++ Rust bindgen story isn't great

**Comment by jdm** [comment]
I believe NSS is pure C as well.

**Comment by frewsxcv** [comment]
Also, for OSX, there's https://github.com/sfackler/rust-security-framework

**Comment by drbo** [comment]
I feel that if we want to reduce our C attack surface, the best answer is to write a pure-Rust crypto library.  Mozilla has cryptographers qualified to do this (they already maintain NSS after all), so the rule of "don't roll your own crypto" doesn't apply here, as long as Mozilla chooses qualified people to do it.

**Comment by metajack** [comment]
FYI, those cryptographers you mention have advised us not to rewrite it.

I believe there is some ongoing work to replace a part of what openssl does with Rust, but until that kind of work is farther along we'll keep using existing solutions.

This isn't to imply that we aren't interested in doing it, but it's just a lower priority for us since we have some much other stuff to do and our team is relatively small. The crypto people have their plates full with their own workload as well.

**Comment by briansmith** [comment]
> FYI, those cryptographers you mention have advised us not to rewrite it.

What was the rationale?

> I believe there is some ongoing work to replace a part of what openssl does with Rust, but until that kind of work is farther along we'll keep using existing solutions.

Interesting. What parts are being replaced?

**Comment by seanmonstar** [comment]
@briansmith I believe he is referring to your work with ring.

**Comment by drbo** [comment]
_ring_  would be my first choice; it uses a subset of BoringSSL.

**Comment by drbo** [comment]
@seanmonstar @briansmith's _ring_ would be my first choice; it uses a subset of BoringSSL (which is the main TLS library used by Google's production servers, and thus expected to be extremely secure) and provides a safe Rust interface.

**Comment by frewsxcv** [comment]
Crate that might help with this: https://github.com/sfackler/rust-native-tls

**Comment by drbo** [comment]
Perhaps BoringSSL can be used instead?  Servo would need to bundle it but it is a much smaller dependency.

**Comment by drbo** [comment]
Note that https://github.com/sfackler/rust-native-tls already includes hostname verification.

**Comment by Ms2ger** [comment]
We definitely don't want platform-specific security code; the plan here is still *ring*, I believe.

**Comment by larsbergstrom** [comment]
In the coming months, we will consult with the Mozilla security and networking teams to get some advice here. Choices of these components are not just based on language alignment, but also organizational (e.g., do we have people on the early/pre-disclosure list to help us have a new release ready for the public announcement? Is there a library that we are contributing to and have high confidence in?).

Further, we don't yet have any security experts on the Servo team, and this is decision has non-trivial impact.

**Comment by nox** [comment]
What's the status on this? Assigning to @avadacatavra to decide.

**Comment by TerraX-net** [comment]
I thought [briansmith/webpki](https://github.com/briansmith/webpki) (with things like [OCSP Stapling](https://github.com/briansmith/webpki/issues/26)/[OCSP Must Staple](https://github.com/briansmith/webpki/issues/28)) and [briansmith/ring](https://github.com/briansmith/ring) are primarily written for Servo because @briansmith aims to have written as much as possible (and reasonable) in Rust?

**Comment by avadacatavra** [comment]
@nox now that we've landed the openssl bump, i'll be putting together a plan to see how we can gradually migrate off openssl

**Comment by DemiMarie** [comment]
@avadacatavra What about ring-tls?

**Comment by djc** [comment]
What's the status of this? As far as I can see the plan of record is to migrate to rustls, but there have been no resources to actually work on the migration?

**Comment by briansmith** [comment]
I think that now that the emphasis of Servo is more clearly not "we must build something equivalent to Firefox/Chrome/etc. ASAP" and more "this is a platform for experimentation" it makes much more sense to switch to Rustls than it did before.

Previously we had a tension about Servo's desire to be compatible with every website that Firefox (et al.) works with, which would have required adding TLS 1.0 and TLS 1.1 support to Rustls and adding some bad cipher suites in order for this to make sense, but basically nobody really wants to add that legacy baggage to Rustls. Because of the re-emphasis of the experimental nature of Servo, clearly Servo doesn't need to work with every website on Earth in the short term so supporting only TLS 1.2 + TLS 1.3 with good cipher suites is perfectly reasonable. The IETF is even working on a draft to kill off TLS 1.0 and TLS 1.1 completely: https://tools.ietf.org/html/draft-moriarty-tls-oldversions-diediedie-00. (Further any PCI-compliant website must be TLS 1.2 already since June 30; see https://www.wired.com/story/tls-encryption-upgrade-credit-card-online-payments/.)

Thus I think this makes a ton of sense to do now. I believe it would be straightforward for even a new contributor who never worked on Servo before to do it.

**Comment by larsbergstrom** [comment]
Our plan for Servo, as mentioned [here](https://blog.servo.org/2018/03/09/servo-and-mixed-reality/) is still to get it out to real end users. This time, hopefully as a separate web engine _alongside_ Gecko that will allow us to experiment with new web platform features that would be infeasible to try out inside of Gecko.

Certainly, per-site experimentation lowers the bar on the requirements for many features such as implementation of the full equivalent web platform before we can ship a first experiment. But, core features that ensure the safety of end users and help to prevent session hijacking are something we'd need to have some really hard conversations about internally before green lighting. That said, I know we're *all* interested in researching areas such as formally verified implementations of secure libraries at some point.

I apologize for the non-answer and continued existence of this open issue, but I hope it helps to know that this is still a conversation we're having internally, and as soon as we have some more clarity around what we feel safe doing in a shipping product, we'll definitely update here with a roadmap!

**Comment by briansmith** [comment]
@larsbergstrom Why isn't the decision-making for this done in the open?

**Comment by larsbergstrom** [comment]
@briansmith Issues like this particular one have direct impact on the larger Mozilla and Firefox organizations, brand, and products and involve gathering information from organizations that do not operate on GitHub in issues. When working with partners or organizations or issues that require closed discussions, we'll do that in order to gather feedback from all affected parties.

This isn't your first time around the block on security stuff - you've been doing it for years longer than I - so I assume you already knew this was the answer :-)

**Comment by briansmith** [comment]
During the time I was at Mozilla there was no need for such secrecy for this kind of decision-making and I don't see the need for it now. Openness of decision-making is a pretty fundamental part of what makes Mozilla Mozilla.

**Comment by briansmith** [comment]
"In March of 2020, Firefox will disable support for TLS 1.0 and TLS 1.1." - https://blog.mozilla.org/security/2018/10/15/removing-old-versions-of-tls/

"In line with these industry standards, Google Chrome will deprecate TLS 1.0 and TLS 1.1 in Chrome 72. Sites using these versions will begin to see deprecation warnings in the DevTools console in that release. TLS 1.0 and 1.1 will be disabled altogether in Chrome 81. This will affect users on early release channels starting January 2020." - https://security.googleblog.com/2018/10/modernizing-transport-security.html

"Today, we’re announcing our intent to disable Transport Layer Security (TLS) 1.0 and 1.1 by default in supported versions of Microsoft Edge and Internet Explorer 11 in the first half of 2020." - https://blogs.windows.com/msedgedev/2018/10/15/modernizing-tls-edge-ie11/

"Therefore, we are deprecating support for TLS 1.0 and 1.1. Complete support will be removed from Safari in updates to Apple iOS and macOS beginning in March 2020." - https://webkit.org/blog/8462/deprecation-of-legacy-tls-1-0-and-1-1-versions/

**Comment by jdm** [comment]
https://github.com/servo/servo/pull/24764 was a PR that added a build switch to use RusTLS instead of openssl. I closed it because the team doesn't want to take on the burden of maintaining additional code without a clear path to making it the default at this time.

**Comment by odedlaz** [comment]
@larsbergstrom, are there any updates regarding the move to [rustls](https://github.com/ctz/rustls)?

**Comment by jdm** [comment]
@odedlaz The last comment is the most up to date news on this subject.

**Comment by odedlaz** [comment]
@jdm, got it. There's a [thread on the mailing list](https://lists.mozilla.org/pipermail/dev-servo/2016-August/001386.html) that listed NSS as a possible candidate to replace OpenSSL. Any ideas why it wasn't picked up?

**Comment by jdm** [comment]
There wasn't a great motivation for the work given the effort it would have required.

**Comment by jdm** [comment]
That being said, https://github.com/mozilla/neqo/tree/master/neqo-crypto now exists and provides Rust bindings for the parts of NSS that are needed by the the QUIC implementation for Firefox. We're still not planning on changing anything right now, but it's good to know that it might be an option in the future.

**Comment by djc** [comment]
Do note that the QUIC interface to TLS is quite different from the interface between TCP and TLS.

**Comment by Darkspirit** [comment]
Hyper's HTTP/3 implementation uses Rustls: https://github.com/hyperium/hyper/pull/1824

So far it turned out useful that NSS did not promote another competing non-Rust dependency in the ecosystem.  Hopefully Mozilla takes over code ownership of Rust's TLS solution (as proposed) and considers steps like creating an NSS API on top of Rustls (if appropriate) - like [Baidu did for OpenSSL](https://github.com/ymjing/tabbyssl) -  to keep [direct control](https://github.com/ctz/webpki-roots#webpki-roots) and [relevance](https://twitter.com/BRIAN_____/status/1217182903899246593).

It would be great if anyone reading this could ping the right people to help funding a [_public_ audit](https://github.com/ctz/rustls/issues/189).

Background on neqo:
After a [suggestion](https://bugzilla.mozilla.org/show_bug.cgi?id=1158011#c15) to switch to Rust's community QUIC implementation which uses Rustls (but could support other backends as well), Mozilla's early [C++ implementation](https://github.com/mcmanus/mozquic) was fortunately abandoned and [Ekr explicitly kept the option open to switch to Rustls at a later point](https://bugzilla.mozilla.org/show_bug.cgi?id=1158011#c19): 
But it seems instead of [directly working with the community](https://bugzilla.mozilla.org/show_bug.cgi?id=1158011#c18) they began working on a [Necko-NSS-proprietary Rust QUIC](https://github.com/mozilla/neqo) implementation as a first step to have _something soon_ so they could avoid cleaning up legacy constraints ([NSPR](https://bugzilla.mozilla.org/show_bug.cgi?id=1158011#c16)) (?).
 Given my AllHands inviter and other good people were still fired in the light of such massive double work, refusal to work with the community and missing participation on Servo network and crypto code, I had cancelled my attendance.

**Comment by Shnatsel** [comment]
A formal audit of rustls has been performed, and found no significant issues. In fact, and I quote,

> Cure53 had the rare pleasure of being incredibly impressed with the presented software.

The audit report can be found here: https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf
